File: comms/torchcomms/tests/integration/cpp/WindowRmaTest.cpp
// Copyright (c) Meta Platforms, Inc. and affiliates.
#include "WindowRmaTest.hpp"

#include <gtest/gtest.h>
#include <ifaddrs.h>
#include <algorithm>
#include "TorchCommTestHelpers.h"

std::unique_ptr<TorchCommTestWrapper> WindowRmaTest::createWrapper() {
  return std::make_unique<TorchCommTestWrapper>();
}

void WindowRmaTest::SetUp() {
  // NCCLX Window RMA requires NCCL_CTRAN_ENABLE=1
  wrapper_ = createWrapper();
  torchcomm_ = wrapper_->getTorchComm();
  rank_ = torchcomm_->getRank();
  num_ranks_ = torchcomm_->getSize();
  device_index_ = rank_ % at::cuda::device_count();
}

void WindowRmaTest::TearDown() {
  // Explicitly reset the TorchComm object to ensure proper cleanup
  torchcomm_.reset();
  wrapper_.reset();
}

bool WindowRmaTest::checkIfSkip() {
  // Check 1: beth0 NIC exists (OSS env requirements)
  struct ifaddrs* ifaddr;
  if (getifaddrs(&ifaddr) == -1) {
    return true; // skip if can't get interface info
  }
  bool nic_found = false;
  for (struct ifaddrs* ifa = ifaddr; ifa != nullptr; ifa = ifa->ifa_next) {
    if (ifa->ifa_name &&
        std::string(ifa->ifa_name).find("beth0") != std::string::npos) {
      nic_found = true;
      break;
    }
  }
  freeifaddrs(ifaddr);
  if (!nic_found) {
    return true; // skip if NIC not found
  }

  // Check 2: NCCL_CTRAN_ENABLE is "1" or "true"
  const char* ctran_env = getenv("NCCL_CTRAN_ENABLE");
  if (!ctran_env) {
    return true; // skip if env not set
  }
  std::string val(ctran_env);
  std::transform(val.begin(), val.end(), val.begin(), ::tolower);
  if (val != "1" && val != "true") {
    return true; // skip if not enabled
  }

  // Check 3: TEST_BACKEND is "ncclx"
  const char* backend_env = getenv("TEST_BACKEND");
  if (!backend_env || std::string(backend_env) != "ncclx") {
    return true; // skip if backend is not ncclx
  }

  // All conditions met, don't skip
  return false;
}

// Test function for basic window allocation & put
void WindowRmaTest::testWindowPutBasic(
    int count,
    at::ScalarType dtype,
    bool async_op,
    bool signal,
    bool async_signal) {
  SCOPED_TRACE(
      ::testing::Message() << "Testing Window Put with count=" << count
                           << " and dtype=" << getDtypeName(dtype));
  // Create tensor with different values based on rank
  at::Tensor input_tensor = createWindowRmaTensor(rank_, count, dtype);
  auto current_stream = at::cuda::getCurrentCUDAStream(device_index_);

  // use mem_pool to allocate the window buffer
  setupMemPool();
  at::Tensor tensor =
      createWindowRmaTensor(rank_, count * num_ranks_, dtype, true);
  auto win = torchcomm_->new_window();
  win->tensor_register(tensor);

  // call barrier to ensure all ranks have allocated the window
  torchcomm_->barrier(false);

  auto dst_rank = (rank_ + 1) % num_ranks_;
  auto src_rank = (rank_ - 1 + num_ranks_) % num_ranks_;

  // Put the tensor to the Window of the next rank using the current stream if
  // async_op is false, otherwise use the internal op_stream
  auto work = win->put(input_tensor, dst_rank, dst_rank * count, async_op);
  if (async_op) {
    // register async op to current stream if async_op is True
    work->wait();
  }

  if (signal) {
    // call signal on current stream to notify remote rank that the put is
    // complete
    auto signal_work = win->signal(dst_rank, async_signal);
    auto wait_signal_work = win->wait_signal(src_rank, async_signal);
    if (async_signal) {
      // register async Signal op to current stream since it is launched on
      // the internal op_stream
      signal_work->wait();
      // register async WaitSignal op to current stream since it is launched on
      // the internal wait_stream
      wait_signal_work->wait();
    }
    // synchronize current stream to ensure signal is complete
    current_stream.synchronize();
  } else {
    // call barrier on current stream to ensure put is complete
    torchcomm_->barrier(false);
  }

  // Get count slices starting from rank_ * count offset
  auto remote_tensor = win->get_tensor(rank_);
  auto result_tensor = remote_tensor.index(
      {at::indexing::Slice(rank_ * count, (rank_ + 1) * count)});

  // Verify results
  verifyWindowRmaResults(result_tensor.cpu(), src_rank);

  win->tensor_deregister();
  // Explicitly destroy the window object to mimic user behavior
  win.reset();
  cleanupMemPool();
}

// Helper function to create tensor for broadcast
at::Tensor WindowRmaTest::createWindowRmaTensor(
    int value,
    int count,
    at::ScalarType dtype,
    bool use_mem_pool) {
  auto options = at::TensorOptions().dtype(dtype).device(device_type_);
  at::Tensor tensor;

  if (use_mem_pool) {
    TORCH_CHECK(
        memPool_ != nullptr,
        "MemPool not initialized. Call setupMemPool() before creating tensors with use_mem_pool=true");
  }

  // Initialize tensor based on dtype
  if (dtype == at::kFloat) {
    tensor = at::ones({count}, options) * static_cast<float>(value);
  } else if (dtype == at::kInt) {
    tensor = at::ones({count}, options) * static_cast<int>(value);
  } else if (dtype == at::kChar) {
    tensor = at::ones({count}, options) * static_cast<signed char>(value);
  }

  return tensor;
}

void WindowRmaTest::setupMemPool() {
  if (memPool_) {
    return; // Already set up
  }

  auto allocator = torchcomm_->getMemAllocator();
  auto cudaAllocator =
      std::dynamic_pointer_cast<c10::cuda::CUDACachingAllocator::CUDAAllocator>(
          allocator);
  memPool_ = std::make_unique<at::cuda::MemPool>(cudaAllocator.get());

  auto tid = std::this_thread::get_id();
  c10::cuda::CUDACachingAllocator::beginAllocateToPool(
      memPool_->device(), memPool_->id(), [=](cudaStream_t) {
        return std::this_thread::get_id() == tid;
      });
}

void WindowRmaTest::cleanupMemPool() {
  if (memPool_) {
    c10::cuda::CUDACachingAllocator::endAllocateToPool(
        memPool_->device(), memPool_->id());
    memPool_.reset();
  }
}

// Helper function to verify results
void WindowRmaTest::verifyWindowRmaResults(
    const at::Tensor& tensor,
    int value) {
  // Use verifyTensorEquality to compare tensor with expected tensor
  std::string description = "Window RMA with value " + std::to_string(value);
  verifyTensorEquality(tensor, value, description);
}

TEST_P(WindowRmaTest, WindowPutBasic) {
  if (checkIfSkip()) {
    GTEST_SKIP() << "Skipping NCCLX-CTRAN-only window tests";
  }
  int count = std::get<0>(GetParam());
  at::ScalarType dtype = std::get<1>(GetParam());
  bool async_op = std::get<2>(GetParam());
  bool signal = std::get<3>(GetParam());
  bool async_signal = std::get<4>(GetParam());
  testWindowPutBasic(count, dtype, async_op, signal, async_signal);
}

INSTANTIATE_TEST_SUITE_P(
    WindowRmaTestParams,
    WindowRmaTest,
    ::testing::Combine(
        // count, dtype, async_op, signal, async_signal
        ::testing::Values(4, 1024, 1024 * 1024),
        ::testing::Values(at::kFloat, at::kInt, at::kChar),
        ::testing::Values(true, false),
        ::testing::Values(true, false),
        ::testing::Values(true, false)),

    [](const ::testing::TestParamInfo<
        std::tuple<int, at::ScalarType, bool, bool, bool>>& info) {
      int count = std::get<0>(info.param);
      at::ScalarType dtype = std::get<1>(info.param);
      std::string async_op = std::get<2>(info.param) ? "asyncOp" : "syncOp";
      std::string signal = std::get<3>(info.param) ? "signal" : "noSignal";
      std::string async_signal =
          std::get<4>(info.param) ? "asyncSignal" : "syncSignal";
      return "Count_" + std::to_string(count) + "_" + getDtypeName(dtype) +
          "_" + async_op + "_" + signal + "_" + async_signal;
    });

// This main function is provided by gtest
int main(int argc, char** argv) {
  ::testing::InitGoogleTest(&argc, argv);
  return RUN_ALL_TESTS();
}

File: comms/torchcomms/tests/integration/py/WindowRmaTest.py
#!/usr/bin/env python3
# pyre-unsafe
# Copyright (c) Meta Platforms, Inc. and affiliates.

import itertools
import os
import unittest

import torch
import torchcomms
from torchcomms import TorchCommWinAccessType
from torchcomms.tests.integration.py.TorchCommTestHelpers import (
    get_dtype_name,
    TorchCommTestWrapper,
)


class WindowRmaTest(unittest.TestCase):
    """Test class for Window RMA operations in TorchComm."""

    def get_wrapper(self):
        return TorchCommTestWrapper()

    def setUp(self):
        self.wrapper = self.get_wrapper()
        self.torchcomm = self.wrapper.get_torchcomm()
        self.rank = self.torchcomm.get_rank()
        self.num_ranks = self.torchcomm.get_size()
        self.device = self.torchcomm.get_device()

        # Get allocator using global function - obtained once and reused
        self.allocator = torchcomms.get_mem_allocator(self.torchcomm.get_backend())

    def tearDown(self):
        self.allocator = None
        self.torchcomm = None
        self.wrapper = None

    def _window_put_test(self, count, dtype, async_op, async_signal):
        """Test window put operation."""
        put_stream = torch.cuda.Stream()
        wait_stream = torch.cuda.Stream()

        # Create input tensor (regular allocation)
        input_tensor = (
            torch.ones(
                [count],
                dtype=dtype,
                device=self.device,
            )
            * self.rank
        )
        # Use the global allocator obtained in setUp
        pool = torch.cuda.MemPool(self.allocator)
        with torch.cuda.use_mem_pool(pool):
            win_buf = torch.ones(
                [count * self.num_ranks], dtype=dtype, device=self.device
            )

        self.torchcomm.barrier(False)

        win = self.torchcomm.new_window()
        win.tensor_register(win_buf)
        self.torchcomm.barrier(False)

        dst_rank = (self.rank + 1) % self.num_ranks
        src_rank = (self.rank - 1 + self.num_ranks) % self.num_ranks

        # Perform multiple put operations to test repeated usage
        num_iterations = 10
        for iteration in range(num_iterations):
            # Put the tensor to the Window of the next rank
            with torch.cuda.stream(put_stream):
                work = win.put(input_tensor, dst_rank, dst_rank * count, async_op)
                if async_op:
                    work.wait()
            # signal the next rank to proceed
            with torch.cuda.stream(put_stream):
                signal_work = win.signal(dst_rank, async_signal)
                if async_signal:
                    signal_work.wait()

            # wait signal from the previous rank
            with torch.cuda.stream(wait_stream):
                wait_signal_work = win.wait_signal(src_rank, async_signal)
                if async_signal:
                    wait_signal_work.wait()

                local_tensor = win.map_remote_tensor(self.rank)

            # wait for the data from the previous rank to be ready
            wait_stream.synchronize()
            output_tensor = local_tensor[self.rank * count : (self.rank + 1) * count]

            target_tensor = (
                torch.ones(
                    [count],
                    dtype=dtype,
                    device=self.device,
                )
                * src_rank
            )

            torch.testing.assert_close(
                output_tensor,
                target_tensor,
                rtol=1e-5,
                atol=1e-5,
                msg=f"Rank {self.rank} Iteration {iteration} Expected {target_tensor} but got {output_tensor}",
            )

        # Cleanup
        wait_stream.synchronize()
        put_stream.synchronize()
        win.tensor_deregister()
        del win
        del pool
        torch.cuda.synchronize()

    def _window_attributes_test(self, count, dtype):
        """Test window attributes for unified vs separate access types."""
        # Get global allocator for the backend and create memory pool
        allocator = torchcomms.get_mem_allocator(self.torchcomm.get_backend())
        pool = torch.cuda.MemPool(allocator)
        with torch.cuda.use_mem_pool(pool):
            win_buf = torch.ones(
                [count * self.num_ranks], dtype=dtype, device=self.device
            )

        win = self.torchcomm.new_window()
        win.tensor_register(win_buf)
        self.torchcomm.barrier(False)

        # Test window attributes
        if self.num_ranks <= 8:
            next_rank = (self.rank + 1) % self.num_ranks
            win_attr = win.get_attr(next_rank)
            assert (
                win_attr.access_type == TorchCommWinAccessType.WIN_ACCESS_TYPE_UNIFIED
            )
        else:
            next_rank = (self.rank + 8) % self.num_ranks
            win_attr = win.get_attr(next_rank)
            assert (
                win_attr.access_type == TorchCommWinAccessType.WIN_ACCESS_TYPE_SEPARATE
            )

        # Cleanup
        win.tensor_deregister()
        del win
        del pool

    def _map_remote_tensor_device_agnostic_test(self, count, dtype):
        """Helper function to test map_remote_tensor with device-agnostic access."""
        print(
            f"Testing map_remote_tensor_device_agnostic with count={count}, dtype={get_dtype_name(dtype)}"
        )

        # Create memory pool and allocate window buffer within it
        pool = torch.cuda.MemPool(self.torchcomm.mem_allocator)
        with torch.cuda.use_mem_pool(pool):
            win_buf = torch.arange(
                count * self.num_ranks, dtype=dtype, device=self.device
            )

        win = self.torchcomm.new_window()
        win.tensor_register(win_buf)
        self.torchcomm.barrier(False)

        # Test local access
        local_tensor = win.map_remote_tensor(self.rank)
        self.assertEqual(local_tensor.dtype, win_buf.dtype)
        self.assertEqual(local_tensor.shape, win_buf.shape)
        torch.testing.assert_close(local_tensor, win_buf, rtol=0, atol=0)

        # Test remote access (only for unified memory)
        remote_rank = (self.rank + 1) % self.num_ranks
        win_attr = win.get_attr(remote_rank)
        if win_attr.access_type == TorchCommWinAccessType.WIN_ACCESS_TYPE_UNIFIED:
            remote_tensor = win.map_remote_tensor(remote_rank)
            self.assertEqual(remote_tensor.dtype, dtype)
            self.assertEqual(remote_tensor.shape, win_buf.shape)

            expected_data = torch.arange(
                count * self.num_ranks, dtype=dtype, device=self.device
            )
            torch.testing.assert_close(
                remote_tensor, expected_data, rtol=1e-5, atol=1e-5
            )

        # Cleanup
        win.tensor_deregister()
        del win
        del pool

    @unittest.skipIf(
        os.getenv("RUN_RMA_TEST", "").lower() not in ("1", "true"),
        "RMA tests require NCCLX backend with CTran enabled (RUN_RMA_TEST=true)",
    )
    def test_all_tests(self):
        """Run all tests with all parameter combinations."""
        counts = [4, 1024, 1024 * 1024]
        dtypes = [torch.float, torch.int, torch.int8]
        async_ops = [True, False]
        async_signals = [True, False]

        # Nested loops for all parameter combinations
        for (
            count,
            dtype,
            async_op,
            async_signal,
        ) in itertools.product(
            counts,
            dtypes,
            async_ops,
            async_signals,
        ):
            # Create a descriptive test name for better test output
            test_name = f"Count_{count}_{get_dtype_name(dtype)}_AsyncOp_{async_op}_AsyncSignal_{async_signal}"
            print(f"Running _window_put_test with parameters: {test_name}")

            self._window_put_test(count, dtype, async_op, async_signal)

        # Test window attributes with specific parameters
        count = 1024
        dtypes_to_test = [torch.float, torch.int, torch.int8]
        for dtype in dtypes_to_test:
            test_name = f"Count_{count}_{get_dtype_name(dtype)}"
            print(f"Running _window_attributes_test with parameters: {test_name}")
            self._window_attributes_test(count, dtype)

        # Test map_remote_tensor_device_agnostic with specific dtypes
        dtypes_to_test = [torch.float32, torch.int32, torch.bfloat16]
        count = 1024
        for dtype in dtypes_to_test:
            print("Running _map_remote_tensor_device_agnostic_test")
            self._map_remote_tensor_device_agnostic_test(count, dtype)


if __name__ == "__main__":
    unittest.main()
